### PR TITLE
Support a silent mode for JWTAuthMechanism

### PR DESCRIPTION
**Pull request by sberyozkin**
Related to #28247.

Hi @FroMage, IMHO this is the best we can do - `JWT` mechanism is not aware of other mechanisms, `Basic+Form` is a special combination, they are built in mechanisms, `vertx-http` knows if both are enabled or not; but also using a silent mode by default appears to be wrong as we can have a case where for example JWT is restricted to a specific path only - which JWT is not aware of - and then returning 302 would be wrong.

So the application which needs it will just enable a `silent` mode for JWT.

The only problem with this PR I just can't get the test passing, because `FormAuthenticationMechanism` (or some custom mechanism - I've tried) is not registered, for some reasons it just does not work in `smallrye-jwt/deployment` and creating a new integration test for it seems a bit too much. 

Can you please try this PR in Renarde ? If it works then I guess we can just commit the main source update only as the change is trivial

### CODE REVIEW

**Comment by FroMage** [comment]
Why not add a `boolean isInteractive()` method to `HttpAuthenticationMechanism` and use it in `HttpAuthenticator.getChallenge`, to figure out if there's any interactive auth in the chain, and then pass it on to the `HttpAuthenticationMechanism.getChallenge` call by adding a parameter which informs the auth mechanism if there's another interactive auth and that makes it silent?

That would have the same effect as the current basic/form combo, and would make it work automatically for jwt/webauthn or any other interactive/header combination, no?

**Comment by sberyozkin** [comment]
Hi @FroMage But the non-interactive mechanism stepping outside and giving a way for the interactive one to challenge is an application specific choice. Even in case of JWT, typically it would be 401, the fact JWT is actually using cookies as in your case to support the already authenticated user is difficult to capture.

We also had a case (discussed in Zulip long time ago - but I can find it),  a user reported a problem, basic and oidc code flow were used, but when no credentials were provided, the user was surprised to get 302 from OIDC, and wanted a Basic challenge. So by default the basic mechanism was sorted in front of OIDC. Surely it was not good for the other user who wanted the opposite 302 by default - so this user uses a path specific configuration to ensure 302 is returned. Hard to generalize it at the Quarkus level such that it works for all.

**Comment by sberyozkin** [comment]
The situation described re Basic and OIDC and 401 being expected actually suggests that having Form preferred by default to create a challenge when no credentials are provided, when both Basic and Form are enabled, may not work for all, Basic Auth challenge done by the browser is a basic interaction with the user...

**Comment by sberyozkin** [comment]
@FroMage Steph, should we start with supporting this config property ? I'm just not sure we can get much progress with the default redirects in case of the interactive mechanisms being present, I'm quite sure that it will break some applications. Perhaps it can be reviewed iteratively

**Comment by sberyozkin** [comment]
For ex, perhaps over time we can make this property optional and doc and dynamically set if it is silent or not depending on some factors, i.e, IMHO having this property won't prevent more improvements going forward

**Comment by FroMage** [comment]
Well, this is not enough, though. It's nice to add the `silent` option. But for users who only use JWT tokens as cookies, how do they redirect to their login page when the auth is missing? This delegates it to webauthn or oidc, but if they only support JWT auth they will have a single auth mechanism.

This is why it feels to me like there should be a single redirect option for all auth mechanisms, so that I can declare that no matter which auth I'm using (JWT, Basic, WebAuthn, OIDC) if auth is required and is missing, redirect to my login page at `/login`.

**Comment by sberyozkin** [comment]
@FroMage 

>This is why it feels to me like there should be a single redirect option for all auth mechanisms, so that I can declare that no matter which auth I'm using (JWT, Basic, WebAuthn, OIDC) if auth is required and is missing, redirect to my login page at /login.

Right, I see the point, but recall I was referring to the concrete case where a user was combining Basic and the interactive Oidc Code flow and was surprised to get 302 instead of a Basic Auth challenge. The point here is that we just can't be 100% sure that the presence of some interactive mechanism means this redirect is expected. And if more than one interactive mechanism is present then Quarkus won't be able to select the correct interactive mechanism anyway without some customization as in case of Renarde for ex.

>Well, this is not enough, though. It's nice to add the silent option.

I agree that it is not a perfect solution, having to manually set this property - but nonetheless it is a way forward. As I said, if we eventually figure out how to make it all automatic such that it works predictably for different users, then we can can just make this property optional and dynamically figure out its value, I just don't see how it can work now...

**Comment by sberyozkin** [comment]
@FroMage It is always useful to have an extension specific way to do something, for example, we have a global trust-all tls setting but while it may work for example for OIDC and may not work for HashiCorp Vault or vice versa etc. Same case here. You have a concrete requirement to have JWT ignore the challenge and have whatever other mechanism which can do it deal with it, and it is supported in this PR.
Global redirect option is a totally different requirement IMHO, redirects should be managed by mechanisms for which redirects make sense.

**Comment by FroMage** [comment]
So you're saying you can fix this in another PR?

**Comment by sberyozkin** [comment]
Hi @FroMage 

>So you're saying you can fix this in another PR?

What I'm saying is that JWT mechanism should not be dealing with redirects, this PR allows to have it skip the challenge if another, interactive, mechanism should be used. Those interactive mechanisms are already configured with redirect URLs, they know how to get the user authenticated via those redirects, so they should be handling it all... 
Making such a behavior a default one seems problematic to me as it is not really possible to guarantee this is what will be expected in every case where JWT is combined with other interactive mechanisms.

This is why having to set `quarkus.smallye.jwt.silent=true` when it is necessary seems a reasonable compromise right now - we just say here, we have some other mechanism which will take care of the challenge. 

While having such an option will not block us in anyway to find a neater solution going forward - but as I said I don't how this a possible follow up solution will look like...
thanks

**Comment by sberyozkin** [comment]
@FroMage Steph, by the way, I've been thinking of your earlier comment,

>Well, OK, so this adds a silent mode, but now I'm pretty sure I'd rather have a global setting for a redirect in case of missing auth, because this would only solve the JWT issue if another extension handles the redirect. If a user doesn't use WebAuthn, or another one which redirects, then this is useless.

But if we have JWT mechanism verifying tokens stored in cookies it only means there is another interactive authentication mechanism which was used originally to authenticate with the authenticated token stored as a cookie. And if no other but JWT mechanisms are used then it means no other mechanism can handle it => 401 is perfect in such a case...
IMHO, we should let this `silent` property in and move on and review whenever possible

**Comment by FroMage** [comment]
> But if we have JWT mechanism verifying tokens stored in cookies it only means there is another interactive authentication mechanism which was used originally to authenticate with the authenticated token stored as a cookie

I do this in my login endpoints, but I've nothing to redirect me to them. That's what's missing from JWT. If I use WebAuthn or OIDC I can configure my redirect. If I only use manual auth and JWT, then I'm out of luck. Perhaps in this case I need to use an exception handler and do the redirect myself.

**Comment by sberyozkin** [comment]
Merging it will help with #30893, use case, jwt and oidc auth mechanisms are used, the new unauthenticated request should start with oidc.
Jwt is restricted to a specific path. So when a jwt auth is required to challenge it will redirect to the oidc controlled path.
Steph,  I appreciate this PR is not solving your reqs, but it can be helpful in simpler cases. Automatic redirects in the presense of more than one user facing mechanism is problematic. Propose to merge it and move on to discussing what extra can be done. 
Cc @FroMage @geoand 
I'll rebase once approved and ask to verify it

**Comment by sberyozkin** [comment]
@kucharzyk Can you please try this PR ? I believe with this PR all you'll need to do is to add the `quarkus.smallrye-jwt.silent=true` configuration, and remove the exception mapper as it will become redundant

**Review by FroMage — APPROVED** [review]

**Comment by FroMage** [comment]
I've decided to go with a replacement of this class to support my use-case: https://github.com/quarkiverse/quarkus-renarde/blob/main/runtime/src/main/java/io/quarkiverse/renarde/util/RenardeJWTAuthMechanism.java

**Comment by sberyozkin** [comment]
Hi @FroMage Sorry, totally missed your approval, I think your replacement is perfect yet simple for the case you are dealing with (and some comments there made me ROFL).
@gsmet It is a very small update that should be safe  to merge to the final, please merge if agreed, otherwise I'll merge after 3.0.0 final is released

**Pull request merged**
